test/libsolidity/syntaxTests/experimental/inference/class_and_instantiation_order.sol

```diff
@@ -0,0 +1,52 @@
+pragma experimental solidity;
+
+type word = __builtin("word");
+
+type T = word;
+
+class Self: C {
+    function f(self: Self) -> word;
+}
+
+instantiation T: C {
+    function f(self: T) -> word {}
+}
+
+instantiation T: D {
+    function f(self: T) -> word {}
+}
+
+class Self: D {
+    function f(self: Self) -> word;
+}
```

Comment:
This is broken after my changes:
```
/solidity/test/soltest.cpp(120): error: in "syntaxTests/experimental/inference/infinite_recursion_two_functions": Exception during extracted test: /solidity/libsolidity/experimental/analysis/TypeInference.cpp(319): Throw in function virtual bool solidity::frontend::experimental::TypeInference::visit(const solidity::frontend::VariableDeclaration &)
Dynamic exception type: boost::wrapexcept<solidity::langutil::InternalCompilerError>
std::exception::what: Solidity assertion failed
[solidity::util::tag_comment*] = Solidity assertion failed
```

Looks like I need a check against multiple visitation in `TypeInference::visit(FunctionDefinition)` after all. The problems is that I can no longer rely on the `type` annotation for that...

Reply:
This *shouldn't* be an issue... this pass shouldn't look at instantiations at all, but only at the class definitions - but at all of those - so the state after this pass should be that all function definitions in class definitions have types from this pass, but none of the ones within intantiations (for types of those we need to rely on inference first)

Reply:
The broken bit is in the type inference pass. I can solve it by adding some extra flag, but I don't like that too much and looks like we'll have more things moved to the extra pass so maybe this problem will just solve itself.